### PR TITLE
ISSUE-9507: ADD `application/gzip,application/octet-stream` accept header

### LINKED ISSUE #9507: Add `accept: application/gzip,application/octet-stream` header when downloading a chart
Hi,

It seems that the current HTTP/HTTPS request made to download a chart from an URL does not specify an `accept` header. This is an issue as some REST APIs serve multiple content types. Since we are specifically requesting a tar.gz (binary) file, the content type should be `application/gzip,application/octet-stream`.

For example, the popular Github API when downloading release binary requires such an `Accept` header, see documentation: https://docs.github.com/en/rest/reference/repos#get-a-release-asset.

This is not limited to Github API but also any other APIs that serve multiple content types.

TLDR add header `accept: application/gzip,application/octet-stream` to HTTPClient when getting charts.

### PR DESCRIPTION
**Pull request by mattclegg**
**What this PR does:**
This PR adds the accept header to the request sent by the HTTPGetter _when downloading archives_.

The function accepts gzip and so it would be preferable if it also notified the server about this. Additionally, this adds support for private helm charts hosted with GitHub.

**Special notes for your reviewer:**
Closes https://github.com/helm/helm/pull/11597
Closes https://github.com/helm/helm/issues/9507

### CODE REVIEW

**Comment by robertsirc** [comment]
So, I will ask the question where are the test for this?

**Comment by mattclegg** [comment]
OK, what would you like to see as a test for this?

- [NewHTTPGetter to contain the accept Header](https://github.com/mattclegg/helm/pull/1)
- [Server with varied Accept Header](https://github.com/mattclegg/helm/pull/2)
- [Ensure headers set return related response](https://github.com/mattclegg/helm/pull/3)

or did you have something else in mind?


You may also test with a private repository on GitHub via;
```
curl -L \
  -H "Authorization: Bearer $TOKEN" \
  -H "X-GitHub-Api-Version: 2022-11-28" \
  https://api.github.com/repos/$OWNER/$REPO/releases/assets/$ASSET_ID
```
which returns;
```
...
< content-type: application/json; charset=utf-8
...
```

but 
```
curl -L \
  -H "Accept: application/octet-stream" \
  -H "Authorization: Bearer $TOKEN" \
  -H "X-GitHub-Api-Version: 2022-11-28" \
  https://api.github.com/repos/$OWNER/$REPO/releases/assets/$ASSET_ID
```
downloads the asset.

**Comment by robertsirc** [comment]
Like in `httpgetter_test` we will expect the useragent and if it isn't correct it throws an error, for example in the test there called `TestDownload` I am questioning if we should have test to ensure a response if we are looking for it in the code. to remain idomatic of the rest of the codebase.

**Comment by mattclegg** [comment]
Thanks @robertsirc
Like this; https://github.com/mattclegg/helm/pull/2/files ?

**Comment by robertsirc** [comment]
@mattclegg yes is this in one PR or two?

**Comment by mattclegg** [comment]
@robertsirc The [other PRs](https://github.com/mattclegg/helm/pulls?q=is%3Apr+is%3Aopen+issue-9507) are to avoid having noise on this branch for other tests that aren't needed.

I've merged the test for the [server with varied Accept Header](https://github.com/helm/helm/pull/13034/commits/866bc05bb29aca832fcee6a911e341996403153c) to this branch. Please let me know if there's anything else that needs to be done before this PR can be merged.

**Comment by mattclegg** [comment]
@robertsirc @mattfarina @gjenkins8 

If you don't mind, could you please check if anything else is needed for this PR?

**Comment by robertsirc** [comment]
Pulled code ran test and my few smoke test ran fine. LGTM.

**Comment by mattfarina** [comment]
@mattclegg I've been testing this with charts in various locations and everything seems to work. But, I tested it alongside the tip of main and I can't reproduce a situation where there would be a problem. Can you provide an example where it would be a problem when using Helm? I understand the curl issue but reproducing with Helm is important, too.

**Comment by mattfarina** [comment]
I was able to reproduce it. It's specific to URLs in the form `https://api.github.com/repos/$OWNER/$REPO/releases/assets/$ASSET_ID` where `$ASSET_ID` is a number..

**Review by mattfarina — APPROVED** [review]

**Comment by mattclegg** [comment]
Would it be possible to get a second approval or any comment from another maintainer, please?

@gjenkins8 @jdolitsky @joejulian @marckhouzam @sabre1041 @scottrigby @technosophos

**Comment by TerryHowe** [comment]
I take it reproducing the problem with $ASSET_ID was more of a ghcr issue than helm.

**Comment by TerryHowe** [comment]
Maybe add 

Closes https://github.com/helm/helm/pull/11597
Closes https://github.com/helm/helm/issues/9507

to your description

**Review by robertsirc — APPROVED** [review]
LGTM

**Pull request merged**
